pkg/controlplane/controller/apiserverleasegc/gc_controller.go

```diff
@@ -0,0 +1,143 @@
+
+	"k8s.io/klog/v2"
+)
+
+const (
+	// Copied from controlplane.IdentityLeaseComponentLabelKey and controlplane.KubeAPIServer
+	// to avoid import cycle.
+	identityLeaseComponentLabelKey = "k8s.io/component"
+	kubeAPIServer                  = "kube-apiserver"
+
+	apiserverIdentityLeaseLabelSelector = identityLeaseComponentLabelKey + "=" + kubeAPIServer
+)
+
+// Controller deletes expired apiserver leases.
+type Controller struct {
+	kubeclientset kubernetes.Interface
+
+	leaseLister   listers.LeaseLister
+	leaseInformer cache.SharedIndexInformer
+	leasesSynced  cache.InformerSynced
+
+	gcCheckPeriod time.Duration
+}
+
+// NewAPIServerLeaseGC creates a new Controller.
+func NewAPIServerLeaseGC(clientset kubernetes.Interface, gcCheckPeriod time.Duration) *Controller {
+	// we construct our own informer because we need such a small subset of the information available.
+	// Just one namespace with label selection.
+	leaseInformer := informers.NewFilteredLeaseInformer(
+		clientset,
+		metav1.NamespaceSystem,
+		12*time.Hour,
+		cache.Indexers{cache.NamespaceIndex: cache.MetaNamespaceIndexFunc},
+		func(listOptions *metav1.ListOptions) {
+			listOptions.LabelSelector = apiserverIdentityLeaseLabelSelector
+		})
+	return &Controller{
+		kubeclientset: clientset,
+		leaseLister:   listers.NewLeaseLister(leaseInformer.GetIndexer()),
+		leaseInformer: leaseInformer,
+		leasesSynced:  leaseInformer.HasSynced,
+		gcCheckPeriod: gcCheckPeriod,
+	}
+}
+
+// Run starts one worker.
+func (c *Controller) Run(stopCh <-chan struct{}) {
+	defer utilruntime.HandleCrash()
+	defer klog.Infof("Shutting down apiserver lease garbage collector")
+
+	klog.Infof("Starting apiserver lease garbage collector")
+
+	// we have a personal informer that is narrowly scoped, start it.
+	go c.leaseInformer.Run(stopCh)
+
+	if !cache.WaitForCacheSync(stopCh, c.leasesSynced) {
+		utilruntime.HandleError(fmt.Errorf("timed out waiting for caches to sync"))
+		return
+	}
+
+	go wait.Until(c.gc, c.gcCheckPeriod, stopCh)
+
+	<-stopCh
+}
+
+func (c *Controller) gc() {
+	selector := labels.NewSelector()
+	r, err := labels.NewRequirement(identityLeaseComponentLabelKey, selection.Equals, []string{kubeAPIServer})
+	if err != nil {
+		klog.Errorf("this should not happen: failed to create apiserver lease label selector: %v", err)
+	}
+	selector = selector.Add(*r)
+
+	leases, err := c.leaseLister.Leases(metav1.NamespaceSystem).List(selector)
+	if err != nil {
+		klog.Errorf("Error while listing apiserver leases: %v", err)
+		return
+	}
+	for _, lease := range leases {
+		// evaluate lease from cache
+		if isLeaseExpired(lease) {
+			// double check latest lease from apiserver before deleting
+			lease, err := c.kubeclientset.CoordinationV1().Leases(metav1.NamespaceSystem).Get(context.TODO(), lease.Name, metav1.GetOptions{})
+			if err != nil && !errors.IsNotFound(err) {
+				klog.Errorf("Error getting lease: %v", err)
+				continue
+			}
+			if errors.IsNotFound(err) || lease == nil {
+				continue
```

Comment:
Is there any component other than this GC expected to delete a lease? If not, let's log a warning.

Reply:
multiple GC controllers can delete the lease simultaneously. Added comments